pennylane/beta/interfaces/autograd.py

```diff
@@ -0,0 +1,164 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the mixin interface class for creating differentiable quantum tapes with
+Autograd.
+"""
+# pylint: disable=protected-access
+import autograd.extend
+import autograd.builtins
+
+from pennylane import numpy as np
+
+from pennylane.beta.queuing import AnnotatedQueue
+
+
+class AutogradInterface(AnnotatedQueue):
+    """Mixin class for applying an autograd interface to a :class:`~.QuantumTape`.
+
+    Autograd-compatible quantum tape classes can be created via subclassing:
+
+    .. code-block:: python
+
+        class MyAutogradQuantumTape(AutogradInterface, QuantumTape):
+
+    Alternatively, the autograd interface can be dynamically applied to existing
+    quantum tapes via the :meth:`~.apply` class method. This modifies the
+    tape **in place**.
+
+    Once created, the autograd interface can be used to perform quantum-classical
+    differentiable programming.
+
+    .. note::
+
+        If using a device that supports native autograd computation and backpropagation, such as
+        :class:`~.DefaultQubitAutograd`, the Autograd interface **does not need to be applied**. It
+        is only applied to tapes executed on non-Autograd compatible devices.
+
+    **Example**
+
+    Once an autograd quantum tape has been created, it can be differentiated using autograd:
+
+    .. code-block:: python
+
+        tape = AutogradInterface.apply(QuantumTape())
+
+        with tape:
+            qml.Rot(0, 0, 0, wires=0)
+            expval(qml.PauliX(0))
+
+        def cost_fn(x, y, z, device):
+            tape.set_parameters([x, y ** 2, y * np.sin(z)], trainable_only=False)
+            return tape.execute(device=device)
+
+    >>> x = np.array(0.1, requires_grad=False)
+    >>> y = np.array(0.2, requires_grad=True)
+    >>> z = np.array(0.3, requires_grad=True)
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> cost_fn(x, y, z, device=dev)
+    [0.03991951]
+    >>> jac_fn = qml.jacobian(cost_fn)
+    >>> jac_fn(x, y, z, device=dev)
+    [[ 0.39828408 -0.00045133]]
+    """
+
+    # pylint: disable=attribute-defined-outside-init
+    dtype = np.float64
+
+    @property
+    def interface(self):  # pylint: disable=missing-function-docstring
+        return "autograd"
+
+    def _update_trainable_params(self):
+        params = []

```

Comment:
This function could use a docstring. I got confused as to its intention, since it both updates a class attribute, but also seems to return important data (which you might consider two distinct actions)

Reply:
Yep, I agree. The reason for the bad practice in this method is optimization.

To get autograd to work, it needs to update the trainable parameters when it accesses the tapes parameters (on call to `get_parameters()`).

So we have:

* `_update_trainable_params()` extracts the tapes parameters, to determine which ones are differentiable
* `get_parameters()` calls `_update_trainable_params()`, and then proceeds to iterate through and extract the trainable parameters _again_.

So you can remove the additional redundant iteration by simply having the data already extract by `_update_trainable_params()` provided to `get_parameters()`.


I don't really like this either, and I tried to improve on it (for example, by changing the name and docstring of `_update_trainable_params` to reflect this behaviour). But, Torch and TensorFlow _don't_ work like this/want this behaviour, so they would then fail. Further, the base `QuantumTape` doesn't need this behaviour either, so _it_ now had a redundant iteration!

I couldn't really think of a better compromise. Perhaps another approach could be to have `AutogradInterface._update_trainable_params()` save a private attribute that only `Autograd.get_parameters()` accesses? This avoids the unexpected return, but trades it off for a hidden side effect. And all in the name of avoiding a redundant for loop 😆

Reply:
perhaps it's just a naming thing. If you called it `function_which_does_x_and_y` I wouldn't be surprised when both x and y happen. Docstring would help as well

Reply:
Rather than renaming the method (which is shared across all interfaces), I both (a) updated the docstring, and (b) changed it to a private attribute:

![image](https://user-images.githubusercontent.com/2959003/93419755-cf2ca000-f8ec-11ea-8a44-40976243ac6f.png)